### ISSUE TITLE
Hooks are executed before .dir-locals are applied

### ISSUE THREAD

**Issue by d12frosted** [syl20bnr/spacemacs#6419]
It seems that hooks are executed before directory locals are applied, so it breaks ability to setup desired tools on per project (directory) basis. For example, setting `ruby-test-runner` to `rspec` in `.dir-locals.el` doesn't enable `rspec-mode`. 

You can also check it by using following content of `.dir-locals.el`.

```
((nil . ((eval . (progn
                   (message " >>>> dir-locals.el"))))))
```

And then execute following elisp code.

```
(defun test-hook-vs-dir-locals ()
  (message " >>>> hook"))

(add-hook #'prog-mode-hook #'test-hook-vs-dir-locals)
```

And then open any file that runs in `prog-mode` derived major mode and observe following output in `*Messages*`.

```
 >>>> hook
 >>>> dir-locals.el
```

At first I thought that there is something different between Emacs 25 and Emacs 24, but this behaviour doesn't change upon switching to Emacs 24.

@syl20bnr do you have any ideas?

**Comment by TheBB** [comment]
Some options in the first answer [here](http://stackoverflow.com/questions/5147060/how-can-i-access-directory-local-variables-in-my-major-mode-hooks). The first one looks quite elegant and easy to implement for us.

**Comment by d12frosted** [comment]
Thanks, I'll take a look. I just thought that it must be working, since @syl20bnr was talking about supporting such flow for haskell layer[1] as it's supported for ruby layer.

[1]: It's pretty much finished, just doesn't work because of this issue

**Comment by TheBB** [comment]
There might be a bug in the ruby layer then. I thought everything that checked the variables in question was postponed until the actual bound functions were called, but I see now that is not the case.

**Comment by syl20bnr** [comment]
I had to use hack-local-variables at some point but I don't remember where, if you search in the code base you should find it. Or maybe it was for elixir when I played with the flycheck checkers but the code may be gone. I'll search for you when I have a chance.

**Comment by syl20bnr** [comment]
It is still here: https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/elixir/funcs.el#L24

The ruby layer should not work indeed... :-(

**Comment by syl20bnr** [comment]
Seems that @TheBB link provides cleaner implementations, since this will be a recurrent pattern it would be cool to have an API to easily dispatch sexps based on dir local vars.

**Comment by d12frosted** [comment]
Solution for Elixir works! I'll test my changes to Haskell layer a little bit and will send PR (probably tomorrow in the morning). I can also fix Ruby layer 😸 

> it would be cool to have an API to easily dispatch sexps based on dir local vars

👍

**Comment by d12frosted** [comment]
Also, current solution is not ideal because it asks for allowance two times. First time because of `(hack-dir-local-variables)` and second time as normal. This is not cool.

**Comment by d12frosted** [comment]
Wow. First solution in that SO answer is so nice and clean. Already using it for haskell layer and I like it so far. Created following simple API:

```
(spacemacs/add-local-vars-hook 'haskell-mode #'spacemacs-haskell//setup-completion-backend)
```

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
